File: caseRouter.js
// caseRouter.js — Router único para pago/pdf/reset
import express from "express";
import PDFDocument from "pdfkit";

/* ====== Config mínima desde ENV ====== */
const RETURN_BASE = process.env.RETURN_BASE || process.env.FRONTEND_BASE || "https://icarticular.cl";
const _ENV = (process.env.KHIPU_ENV || "integration").toLowerCase();
const KHIPU_MODE =
  _ENV === "guest" ? "guest" :
  (_ENV === "prod" || _ENV === "production") ? "production" : "integration";

const KHIPU_API_BASE = "https://payment-api.khipu.com";
const KHIPU_API_KEY  = process.env.KHIPU_API_KEY || "";
const KHIPU_AMOUNT   = Number(process.env.KHIPU_AMOUNT || 1000);
const KHIPU_SUBJECT  = process.env.KHIPU_SUBJECT || "Orden médica ICA";
const CURRENCY       = "CLP";

/* ====== Utils ====== */
const ns = (s, id) => `${s}:${id}`;
const sanitize = (t) => String(t || "").replace(/[^a-zA-Z0-9_-]+/g, "_");
const normRut = (s="") => String(s).replace(/[^0-9kK]/g,"").toUpperCase();
const getBackendBase = (req) => `${req.protocol}://${req.get("host")}`;
const getCase = (memoria, space, id) => memoria.get?.(ns(space, id)) || null;

function mergeNoDestructivo(prev = {}, incoming = {}) {
  const next = { ...prev };
  for (const [k, v] of Object.entries(incoming)) {
    if (v === undefined) continue;
    if (typeof v === "string" && v.trim() === "") continue;
    if (Array.isArray(v) && v.length === 0) continue;
    next[k] = v;
  }
  // preservar campos críticos
  if (Array.isArray(prev.examenesIA) && (!Array.isArray(next.examenesIA) || next.examenesIA.length === 0)) {
    next.examenesIA = prev.examenesIA;
  }
  if (prev.diagnosticoIA && !next.diagnosticoIA) next.diagnosticoIA = prev.diagnosticoIA;
  if (prev.justificacionIA && !next.justificacionIA) next.justificacionIA = prev.justificacionIA;
  if (prev.rmForm && !next.rmForm) next.rmForm = prev.rmForm;
  if (prev.rmObservaciones && !next.rmObservaciones) next.rmObservaciones = prev.rmObservaciones;
  return next;
}

function limpiarCaso(memoria, id) {
  const spaces = ["ia","trauma","preop","generales","meta"];
  let removed = 0;
  for (const s of spaces) removed += memoria.delete(ns(s, id)) ? 1 : 0;
  return removed;
}

/* ====== Lazy loaders de generadores PDF ====== */
let _genTrauma = null;
async function loadOrdenImagenologia(){ if(_genTrauma) return _genTrauma; const m = await import("./ordenImagenologia.js"); _genTrauma = m.generarOrdenImagenologia; return _genTrauma; }
let _genPreopLab=null,_genPreopOd=null;
async function loadPreop(){ if(!_genPreopLab){const m=await import("./preopOrdenLab.js"); _genPreopLab=m.generarOrdenPreopLab;} if(!_genPreopOd){const m=await import("./preopOdonto.js"); _genPreopOd=m.generarPreopOdonto;} return {_genPreopLab,_genPreopOd}; }
let _genGenerales=null;
async function loadGenerales(){ if(_genGenerales) return _genGenerales; const m=await import("./generalesOrden.js"); _genGenerales=m.generarOrdenGenerales; return _genGenerales; }
let _genRM=null;
async function loadFormularioRM(){ if(_genRM) return _genRM; const m = await import("./resonanciaFormularioPDF.js"); _genRM = m.generarFormularioResonancia; return _genRM; }

/* ====== Texto para PDF (reuse de tus helpers) ====== */
function buildExamenTextoStrict(rec={}) {
  if (Array.isArray(rec.examenesIA) && rec.examenesIA.length > 0) {
    return rec.examenesIA.map(x => String(x||"").trim()).filter(Boolean).join("\n");
  }
  if (typeof rec.examen === "string" && rec.examen.trim()) return rec.examen.trim();
  return "";
}
function buildNotaStrict(rec={}) {
  if (typeof rec.nota === "string" && rec.nota.trim()) return rec.nota.trim();
  if (typeof rec.observaciones === "string" && rec.observaciones.trim()) return rec.observaciones.trim();
  if (typeof rec.informeIA === "string" && rec.informeIA.trim()) return rec.informeIA.trim();
  return "";
}
function contieneRM(texto=""){ const s=String(texto||"").toLowerCase(); return s.includes("resonancia") || s.includes("resonancia magn") || /\brm\b/i.test(texto); }

/* ====== Router ====== */
export default function caseRouter(app){
  const router = express.Router();
  const memoria = app.get("memoria");

  // PAGO (central, incluye guest)
  router.post("/pay", async (req, res) => {
    try {
      const { idPago, modulo, datosPaciente, modoGuest } = req.body || {};
      if (!idPago) return res.status(400).json({ ok:false, error:"Falta idPago" });

      const space = (modulo || "").toLowerCase() === "preop" ? "preop"
                   : (modulo || "").toLowerCase() === "generales" ? "generales"
                   : (modulo || "").toLowerCase() === "ia" ? "ia"
                   : "trauma";

      // merge data
      if (datosPaciente) {
        const prev = getCase(memoria, space, idPago) || {};
        const next = mergeNoDestructivo(prev, datosPaciente);
        memoria.set(ns(space, idPago), next);
      }
      memoria.set(ns("meta", idPago), { moduloAutorizado: space });

      // GUEST flow
      if (modoGuest === true || KHIPU_MODE === "guest") {
        if (space === "ia") { // IA exige pagoConfirmado para su PDF
          const prev = getCase(memoria, space, idPago) || {};
          memoria.set(ns(space, idPago), { ...prev, pagoConfirmado: true });
        }
        const url = new URL(RETURN_BASE);
        url.searchParams.set("pago","ok");
        url.searchParams.set("idPago", idPago);
        url.searchParams.set("modulo", space);
        return res.json({ ok:true, url: url.toString(), guest:true });
      }

      // REAL Khipu
      if (!KHIPU_API_KEY) return res.status(500).json({ ok:false, error:"Falta KHIPU_API_KEY" });

      const backendBase = getBackendBase(req);
      const payload = {
        amount: KHIPU_AMOUNT,
        currency: CURRENCY,
        subject: KHIPU_SUBJECT,
        transaction_id: idPago,
        return_url: `${RETURN_BASE}?pago=ok&idPago=${encodeURIComponent(idPago)}&modulo=${space}`,
        cancel_url: `${RETURN_BASE}?pago=cancelado&idPago=${encodeURIComponent(idPago)}&modulo=${space}`,
        notify_url: `${backendBase}/webhook`,
      };

      const r = await fetch(`${KHIPU_API_BASE}/v3/payments`, {
        method: "POST",
        headers: { "content-type":"application/json", "x-api-key": KHIPU_API_KEY },
        body: JSON.stringify(payload),
      });

      const j = await r.json().catch(()=> ({}));
      if (!r.ok) {
        const msg = j?.message || `Error Khipu (${r.status})`;
        return res.status(502).json({ ok:false, error: msg, detail: j || null });
      }
      const urlPago = j?.payment_url || j?.simplified_transfer_url || j?.url;
      if (!urlPago) return res.status(502).json({ ok:false, error:"Khipu no entregó payment_url", detail:j });

      return res.json({ ok:true, url: urlPago });
    } catch (e) {
      console.error("/case/pay error:", e);
      return res.status(500).json({ ok:false, error:e.message });
    }
  });

  // PDF unificado (con reset opcional)
  router.get("/pdf/:idPago", async (req, res) => {
    try {
      const id = req.params.idPago;
      const modulo = String(req.query.modulo || "").toLowerCase();
      const reset = String(req.query.reset || "0") === "1";

      const meta = memoria.get(ns("meta", id));
      const space = modulo || meta?.moduloAutorizado;
      if (!space) return res.status(400).json({ ok:false, error:"Falta modulo" });
      if (!meta || meta.moduloAutorizado !== space) return res.sendStatus(402);

      let d = getCase(memoria, space, id);
      if (!d) return res.sendStatus(404);

      // IA exige pago confirmado
      if (space === "ia" && !d.pagoConfirmado) return res.sendStatus(402);

      let generar, filename, writer;
      const rut = normRut(d.rut || d.RUT || d.RUN || "");
      const basePDF = () => {
        const doc = new PDFDocument({ size:"A4", margin:50 });
        doc.pipe(res);
        return doc;
      };

      // Config por módulo
      if (space === "trauma") {
        generar = await loadOrdenImagenologia();
        const datos = { ...d, examen: buildExamenTextoStrict(d), nota: buildNotaStrict(d), rut };
        filename = `orden_${sanitize(d.nombre || "paciente")}.pdf`;
        res.setHeader("Content-Type", "application/pdf");
        res.setHeader("Content-Disposition", `attachment; filename="${filename}"`);
        writer = basePDF(); generar(writer, datos); writer.end();
      } else if (space === "preop") {
        const { _genPreopLab, _genPreopOd } = await loadPreop();
        filename = `preop_${sanitize(d.nombre || "paciente")}.pdf`;
        res.setHeader("Content-Type", "application/pdf");
        res.setHeader("Content-Disposition", `attachment; filename="${filename}"`);
        writer = basePDF(); _genPreopLab(writer, { ...d, rut }); writer.addPage(); _genPreopOd(writer, { ...d, rut }); writer.end();
      } else if (space === "generales") {
        generar = await loadGenerales();
        filename = `generales_${sanitize(d.nombre || "paciente")}.pdf`;
        res.setHeader("Content-Type", "application/pdf");
        res.setHeader("Content-Disposition", `attachment; filename="${filename}"`);
        writer = basePDF(); generar(writer, { ...d, rut }); writer.end();
      } else if (space === "ia") {
        generar = await loadOrdenImagenologia();
        const datos = { ...d, examen: buildExamenTextoStrict(d), nota: buildNotaStrict(d), rut };
        filename = `ordenIA_${sanitize(d.nombre || "paciente")}.pdf`;
        res.setHeader("Content-Type", "application/pdf");
        res.setHeader("Content-Disposition", `attachment; filename="${filename}"`);
        writer = basePDF(); generar(writer, datos); writer.end();
      } else {
        return res.status(400).json({ ok:false, error:"Módulo inválido" });
      }

      if (reset) {
        res.on("finish", () => {
          try { limpiarCaso(memoria, id); } catch (e) { console.warn("reset post-PDF falló:", e); }
        });
      }
    } catch (e) {
      console.error("/case/pdf error:", e);
      res.sendStatus(500);
    }
  });

  // RESET unificado
  router.delete("/:idPago", (req,res) => {
    try {
      const id = req.params.idPago;
      if (!id) return res.status(400).json({ ok:false, error:"Falta idPago" });
      const meta = memoria.get(ns("meta", id));
      if (!meta) return res.status(404).json({ ok:false, error:"Caso no encontrado o no autorizado" });
      const removed = limpiarCaso(memoria, id);
      return res.json({ ok:true, removed });
    } catch (e) {
      console.error("/case delete error:", e);
      return res.status(500).json({ ok:false, error:"No se pudo limpiar el caso" });
    }
  });

  return router;
}

File: routers/case_router.py
# routers/case_router.py
# Router unificado: pago / PDF / reset
# Stateless — sin memoria server-side
# Depende de: FastAPI, httpx

import logging
import re
from urllib.parse import urlencode

import httpx
from fastapi import APIRouter, HTTPException, Request, Response
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from typing import Any

logger = logging.getLogger("case_router")

# ============================================================
# CONFIG (se inyecta desde main.py via app.state)
# ============================================================
# Acceso: request.app.state.config

KHIPU_API_BASE = "https://payment-api.khipu.com"
CURRENCY       = "CLP"

ESPACIOS = ["trauma", "preop", "generales", "ia"]

router = APIRouter(prefix="/case", tags=["case"])


# ============================================================
# UTILS
# ============================================================
def _sanitize(t: str) -> str:
    return re.sub(r'[^a-zA-Z0-9_-]+', '_', str(t or ""))


def _norm_rut(s: str) -> str:
    return re.sub(r'[^0-9kK]', '', str(s or "")).upper()


def _build_examen_texto(rec: dict) -> str:
    examenes = rec.get("examenes_ia") or rec.get("examenesIA") or []
    if isinstance(examenes, list) and examenes:
        return "\n".join(str(x).strip() for x in examenes if str(x).strip())
    examen = rec.get("examen") or ""
    return str(examen).strip()


def _build_nota(rec: dict) -> str:
    for campo in ["nota", "observaciones", "informe_ia", "informeIA"]:
        v = rec.get(campo)
        if isinstance(v, str) and v.strip():
            return v.strip()
    return ""


def _modulo_desde_body(modulo: str) -> str:
    m = str(modulo or "").lower()
    if m == "preop":    return "preop"
    if m == "generales": return "generales"
    if m == "ia":       return "ia"
    return "trauma"


# ============================================================
# SCHEMAS
# ============================================================
class PayBody(BaseModel):
    idPago:        str
    modulo:        str | None = None
    datosPaciente: dict | None = None
    modoGuest:     bool = False


# ============================================================
# ENDPOINT: POST /case/pay
# ============================================================
@router.post("/pay")
async def pay(body: PayBody, request: Request):
    cfg = request.app.state.config
    id_pago = body.idPago
    space   = _modulo_desde_body(body.modulo or "")

    khipu_mode   = str(cfg.get("khipu_env") or "integration").lower()
    khipu_mode   = "production" if khipu_mode in ("prod", "production") else \
                   "guest"      if khipu_mode == "guest" else "integration"
    return_base  = cfg.get("return_base") or cfg.get("frontend_base") or "https://icarticular.cl"
    khipu_key    = cfg.get("khipu_api_key") or ""
    khipu_amount = int(cfg.get("khipu_amount") or 1000)
    khipu_subject = cfg.get("khipu_subject") or "Orden médica ICA"

    # GUEST flow
    if body.modoGuest or khipu_mode == "guest":
        params = urlencode({"pago": "ok", "idPago": id_pago, "modulo": space})
        url = f"{return_base}?{params}"
        return {"ok": True, "url": url, "guest": True}

    # REAL Khipu
    if not khipu_key:
        raise HTTPException(500, detail="Falta KHIPU_API_KEY")

    base_url = str(request.base_url).rstrip("/")
    payload = {
        "amount":         khipu_amount,
        "currency":       CURRENCY,
        "subject":        khipu_subject,
        "transaction_id": id_pago,
        "return_url":     f"{return_base}?pago=ok&idPago={id_pago}&modulo={space}",
        "cancel_url":     f"{return_base}?pago=cancelado&idPago={id_pago}&modulo={space}",
        "notify_url":     f"{base_url}/webhook",
    }

    try:
        async with httpx.AsyncClient(timeout=15.0) as client:
            r = await client.post(
                f"{KHIPU_API_BASE}/v3/payments",
                headers={
                    "content-type": "application/json",
                    "x-api-key":    khipu_key,
                },
                json=payload,
            )
        j = r.json() if r.content else {}
        if not r.is_success:
            msg = j.get("message") or f"Error Khipu ({r.status_code})"
            raise HTTPException(502, detail=msg)

        url_pago = j.get("payment_url") or j.get("simplified_transfer_url") or j.get("url")
        if not url_pago:
            raise HTTPException(502, detail="Khipu no entregó payment_url")

        return {"ok": True, "url": url_pago}

    except HTTPException:
        raise
    except Exception as e:
        logger.error("/case/pay error: %s", e)
        raise HTTPException(500, detail=str(e))


# ============================================================
# ENDPOINT: GET /case/pdf/{id_pago}
# ============================================================
@router.get("/pdf/{id_pago}")
async def pdf(id_pago: str, request: Request, modulo: str = "", reset: int = 0):
    from ordenes.orden_imagenologia    import generar_orden_imagenologia
    from ordenes.ia_orden_imagenologia import generar_orden_imagenologia_ia
    from ordenes.preop_orden_lab       import generar_orden_preop_lab
    from ordenes.preop_odonto          import generar_preop_odonto
    from ordenes.generales_orden       import generar_orden_generales

    cfg     = request.app.state.config
    space   = _modulo_desde_body(modulo) if modulo else ""

    if not space:
        raise HTTPException(400, detail="Falta modulo")

    # En modo stateless los datos deben venir en query o el router los tiene
    # Este endpoint requiere que el frontend provea los datos via POST /case/pdf
    # — ver nota abajo. Por ahora devuelve 501 si no hay datos en state.
    # (Se completa en main.py con el flujo real de datos)
    raise HTTPException(501, detail=(
        "Endpoint /case/pdf requiere datos del paciente. "
        "Usar POST /case/pdf con body completo."
    ))


# ============================================================
# ENDPOINT: POST /case/pdf  (stateless — datos en body)
# ============================================================
class PdfBody(BaseModel):
    idPago:  str
    modulo:  str
    datos:   dict
    reset:   bool = False


@router.post("/pdf")
async def pdf_post(body: PdfBody, request: Request):
    from ordenes.orden_imagenologia    import generar_orden_imagenologia
    from ordenes.ia_orden_imagenologia import generar_orden_imagenologia_ia
    from ordenes.preop_orden_lab       import generar_orden_preop_lab
    from ordenes.preop_odonto          import generar_preop_odonto
    from ordenes.generales_orden       import generar_orden_generales
    from pdf.generador                 import generar_informe_ia

    space  = _modulo_desde_body(body.modulo)
    d      = body.datos
    rut    = _norm_rut(d.get("rut") or d.get("RUT") or d.get("RUN") or "")
    nombre = _sanitize(d.get("nombre") or "paciente")

    try:
        if space == "trauma":
            pdf_bytes = generar_orden_imagenologia({
                **d,
                "examen": _build_examen_texto(d),
                "nota":   _build_nota(d),
                "rut":    rut,
            })
            filename = f"orden_{nombre}.pdf"

        elif space == "preop":
            # Dos páginas: lab + odonto
            import io
            from reportlab.lib.pagesizes import A4
            from reportlab.platypus import SimpleDocTemplate, PageBreak
            # Generamos cada PDF por separado y los concatenamos
            from pypdf import PdfWriter, PdfReader
            lab_bytes   = generar_orden_preop_lab({**d, "rut": rut})
            odonto_bytes = generar_preop_odonto({**d, "rut": rut})

            writer = PdfWriter()
            for parte in [lab_bytes, odonto_bytes]:
                reader = PdfReader(io.BytesIO(parte))
                for page in reader.pages:
                    writer.add_page(page)
            out = io.BytesIO()
            writer.write(out)
            pdf_bytes = out.getvalue()
            filename = f"preop_{nombre}.pdf"

        elif space == "generales":
            pdf_bytes = generar_orden_generales({
                **d,
                "examenes_ia": d.get("examenes_ia") or d.get("examenesIA") or [],
                "rut": rut,
            })
            filename = f"generales_{nombre}.pdf"

        elif space == "ia":
            pdf_bytes = generar_informe_ia({
                **d,
                "examen": _build_examen_texto(d),
                "nota":   _build_nota(d),
                "rut":    rut,
            })
            filename = f"ordenIA_{nombre}.pdf"

        else:
            raise HTTPException(400, detail="Módulo inválido")

    except HTTPException:
        raise
    except Exception as e:
        logger.error("/case/pdf error: %s", e)
        raise HTTPException(500, detail=str(e))

    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


# ============================================================
# ENDPOINT: DELETE /case/{id_pago}  (reset — sin-op en stateless)
# ============================================================
@router.delete("/{id_pago}")
async def reset(id_pago: str):
    # Stateless: no hay memoria que limpiar
    # El frontend limpia sessionStorage por su cuenta
    logger.info("Reset solicitado para idPago=%s (stateless: no-op)", id_pago)
    return {"ok": True, "removed": 0}
  
